// generator-xtc/module/index.js
'use strict';
var util = require('util');
var path = require('path');
var fs = require('fs');
var yeoman = require('yeoman-generator');
var chalk = require('chalk');
var fullname = require('fullname');
var log = console.log;


var XtcGenerator = yeoman.generators.Base.extend({

	constructor: function() {
		yeoman.generators.Base.apply(this, arguments);

		this.argument('name' , {
			desc: 'The name of the new module'
			,required: false
			,type: String
		});
	},


	initializing: function () {
		this.xtcPath = this.options.path ? path.resolve(process.cwd(), this.options.path) : path.join(process.cwd(), 'node_modules/xtc');
		this.xtcCfg = require(path.join(this.xtcPath, 'lib/configure.js')).getRaw();
		this.projectPath = process.cwd();

		// welcome message
		// http://patorjk.com/software/taag/#p=display&f=Pepper&t=xtc%20mod%20gen
		var welcome =
			'\n' +
			chalk.yellow.bold("      _/__   _ _  _   _/  _  _  _ ") + '\n' +
			chalk.yellow.bold("    ></ /_  / / //_//_/  /_//_'/ /") + '\n' +
			chalk.yellow.bold("                         _/       ") + '\n\n' +
			chalk.blue.bold('    express-terrific module generator\n\n') +
			chalk.green('Please answer a few questions to create your new module.\n')
		;
		this.log(welcome);
	},


	moduleName: function() {
		var cb = this.async();

		// module name
		var prompts = [
			{
				name    : 'name'
				,message: 'What\'s the name of the new module?\n' +
				          '(all lowercase, hyphen separated, without any prefix)'
				,default: this.name
				,validate: validateString
			}
		];

		this.prompt(prompts, function(props) {
			this.modName = props.name.toLowerCase();
			cb();
		}.bind(this));
	},


	customize: function() {
		var cb = this.async();

		nl();

		// accept defaults?
		var prompts = [
			{
				type: 'confirm'
				,name: 'acceptDefaults'
				,message: 'By default this module will have one template, a style sheet, a JS module (with tests) and a README.\n' +
				          'Ok?'
				,default: true
			}
		];

		this.prompt(prompts, function (props) {
			this.customize = !props.acceptDefaults;

			cb();
		}.bind(this));
	},


	doCustomize: function() {
		var cb = this.async();

		if (!this.customize) {
			cb();
			return;
		}

		nl();

		var prompts = [
			{
				type: 'list'
				,name: 'customizeSelection'
				,message: 'Please choose'
				,choices: [
					{
						 name: 'JS only'
						,value: ['needJs']
					}
					,{
						 name: 'HTML only'
						,value: ['needHtml']
					}
					,{
						 name: 'HTML + CSS'
						,value: ['needHtml', 'needCss']
					}
					,{
						 name: 'HTML + JS'
						,value: ['needHtml', 'needJs']
					}
					,{
						 name: 'JS + CSS'
						,value: ['needJs', 'needCss']
					}
					,{
						 name: 'HTML + CSS + JS'
						,value: ['needHtml', 'needCss', 'needJs']
					}
				]
			}
		];

		this.prompt(prompts, function (answer) {

			answer.customizeSelection.forEach(function(value, index, array) {
				this[value] = true;
			}, this)

			cb();
		}.bind(this));
	},


	configure: function() {
		var done = this.async();

		this.nameCSS = 'mod-' + this.modName;
		this.nameFolder = this.xtcCfg.moduleDirName.replace('{{name}}', this.modName);
		this.nameJS = toCamel(this.nameCSS).replace('mod', '');
		this.nameTest = this.name + '.test.js';
		this.modulesDir = path.join(this.xtcCfg.sources.modulesBaseDir, this.nameFolder);

		fullname((function (err, name) {
			if (err) this.user = process.env.USER;
			this.user = name;
			done();
		}).bind(this));
	},


	files: function() {
		nl();
		if (!this.customize || this.needHtml) {
			if (this.xtcCfg.customModuleTemplatePath) {
				// Project defines custom module template(s)
				var stat = fs.lstatSync(path.resolve(this.projectPath, this.xtcCfg.customModuleTemplatePath));
				if (stat.isFile()) {
					this.mkdir(this.modulesDir);
					this.copy(path.resolve(this.projectPath, this.xtcCfg.customModuleTemplatePath), path.join(this.modulesDir, this.modName + (this.xtcCfg.templateExtension || '.hbs' )));
				}
				else if (stat.isDirectory()) {
					this.directory(path.resolve(this.projectPath, this.xtcCfg.customModuleTemplatePath), path.join(this.modulesDir));
				}
				else {
					this.log(chalk.red('The custom module template can\'t be found: %s', this.xtcCfg.customModuleTemplatePath));
					process.exit(1);
				}
			}
			else {
				// Use default module template
				this.mkdir(this.modulesDir);
				this.template('_module.hbs', path.join(this.modulesDir, this.modName + (this.xtcCfg.templateExtension || '.hbs' )));
			}
		}
		(!this.customize || this.needCss) && this.template('_module.less', path.join(this.modulesDir, this.modName +'.less'));

		if (!this.customize || this.needJs) {
			this.template('_module.js', path.join(this.modulesDir, this.modName +'.js'));
			this.mkdir(path.join(this.modulesDir, 'test'));
			this.template('test/_module.test.js', path.join(this.modulesDir, 'test', this.nameTest));
		}
		this.template('_README.md', path.join(this.modulesDir, 'README.md'));
	},


	end: function() {

		if (this.xtcCfg.templateExtension === '.hbs') {
			this.log('\nSnippet:  '+ chalk.blue.bold('{{mod "%s"}}') +'\n', this.modName);
		}
	}


});


module.exports = XtcGenerator;



/**
* Camelizes the given string.
*
* @method toCamel
* @param {String} str
* The original string
* @return {String}
* The camelized string
*/
var toCamel = function(str){
	return str.replace(/(\-[A-Za-z])/g, function($1){return $1.toUpperCase().replace('-','');});
}

function nl(count) {
	count = count || 1;

	while (count > 0) {
		console.log('');
		count--;
	}
}

function validateString(value) {
	return typeof value === 'string' && value.length > 0;
}
