// generator-xtc/app/index.js
'use strict';

var util = require('util');
var path = require('path');
var fs = require('fs');
var yeoman = require('yeoman-generator');
var chalk = require('chalk');
var _projectPath = '';
var log = console.log;
var dir = console.dir;


var XtcGenerator = module.exports = function XtcGenerator(args, options, config) {
	yeoman.generators.Base.apply(this, arguments);

	this.projectPath = process.cwd();
	this.xtcPath = path.join(process.cwd(), 'node_modules/xtc');
	var xtcPkg = require(path.join(this.xtcPath, 'package.json'));
	this.xtcVersion = xtcPkg.version;
	_projectPath = this.projectPath;

	// .xtcrc file stores setup config for setting defaults when running setup again or updating
	var getRc = function(projectPath) {
		var fs = require('fs')
		   ,rcData = {
				 rcVersion  : 0
				,xtcVersion : xtcPkg.version
			}
		;

		try {
			rcData = fs.readFileSync(path.join(projectPath, '/.xtcrc'));
			rcData = JSON.parse(rcData);
		} catch (e) {
			if (e.code !== 'ENOENT') {
				console.error(e.message);
			}
		}
		return rcData;
	};
	this.rc = getRc(_projectPath);


	this.on('end', onEnd);

	/*this.on('end', function () {
		this.installDependencies({
			skipInstall: options['skip-install'],
			callback: onDependenciesInstalled.bind(this)
		});
	});*/

	var intro =
		chalk.cyan('\nxtc@%s\n\n') +
		chalk.magenta("    _/__    _  __  ._  _ _/_  _  _  _ ") + '\n' +
		chalk.magenta("  ></ /_   /_///_///_'/_ /   /_//_'/ /") + '\n' +
		chalk.magenta("          /     |/           _/       ") + '\n\n' +
		chalk.blue(' * express-terrific project generator *\n\n') +
		chalk.cyan('Please answer a few questions to set up your new project.\n')
	;

	log(intro, this.xtcVersion);
};

util.inherits(XtcGenerator, yeoman.generators.Base);


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// NAME AND REPOSITORY


XtcGenerator.prototype.basics = function basics() {
	var cb = this.async()
		,prompts = [
		{
			name: 'name'
			,message: 'Enter a package name for the new project (no spaces)'
			,validate: validateString
			,default: this.rc.name
		},
		{
			name: 'repoUri'
			,message: '"Read-only" repository URI (https://host/path/to/project.git or blank)'
			,validate: validateGitUri
			,default: this.rc.repositoryUri
		}
	];

	this.repository = {};

	this.prompt(prompts, function(props) {
		this.name = props.name;
		this.repository.uri = props.repoUri;
		cb();
	}.bind(this));
};


XtcGenerator.prototype.gitBranch = function gitBranch() {
	var cb, prompts;

	if (!this.repository.uri) { return }
	cb = this.async();

	prompts = [
		{
			name: 'branch'
			,message: 'What branch should be used when linking to project source files'
			,default: this.rc.repositoryBranch || 'develop'
		}
	];

	this.prompt(prompts, function(props) {
		this.repository.branch = props.branch;
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// CHOOSE SERVER AND BUILD PARTS


XtcGenerator.prototype.parts = function parts() {
	var cb = this.async()
		,prompts = [
		{
			name: 'neededParts'
			,type: 'checkbox'
			,message: 'What parts of xtc do you need'
			,choices: [
				{
					name: 'Server'
					,value: 'needServer'
					,checked: typeof this.rc.needServer == 'boolean' ? this.rc.needServer : true
				}
				,{
					name: 'Build'
					,value: 'needBuild'
					,checked: typeof this.rc.needBuild == 'boolean' ? this.rc.needBuild : true
				}
			]
		}
	];

	this.needBuild = false;
	this.needServer = false;
	this.configs = ['"default"'];

	this.prompt(prompts, function(props) {
		var i;
		for (i = 0; i < props.neededParts.length; i++) {
			var value = props.neededParts[i];
			this[value] = true;
		}

		this.needBuild && this.configs.push('"build"');
		this.needServer && this.configs.push('"server"', '"secret"');
		this.configs.push('"local"');
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// BUILD CONFIG


XtcGenerator.prototype.glue = function glue() {
	var cb, prompts
		,_projectPath = this.projectPath
		,isDir = false
		,templateExtension = '.hbs'
	;

	if (!this.needBuild) { return }
	cb = this.async();

	prompts = [
		{
			name    : 'enableSprites'
			,type   : 'confirm'
			,message: 'Enable sprites generation? Note: Needs a working "glue" (Python) installation.'
			,default: typeof this.rc.enableSprites == 'boolean' ? this.rc.enableSprites : false
		}
		,{
			name    : 'customModuleTemplate'
			,type   : 'confirm'
			,message: 'Do you need to use a custom template when generating modules?'
			,default: typeof this.rc.customModuleTemplatePath == 'string'
		}
		,{
			name    : 'customModuleTemplatePath'
			,when   : function(answers) { return answers.customModuleTemplate }
			,type   : 'input'
			,message: 'Template for module generator, can be file or directory'
			,default: this.rc.customModuleTemplatePath
			,validate: function(templatePath) {

				try {
					var stat = fs.lstatSync(path.resolve(_projectPath, templatePath));
				}
				catch (e) {
					return 'Path does not exist';
				}

				if (stat.isFile()) {
					templateExtension = path.extname(templatePath);
					return true;
				}
				if (stat.isDirectory()) {
					isDir = true;
					return true;
				}
			}
		}
		,{
			name    : 'templateExtension'
			,when   : function(answers) { return isDir }
			,type   : 'input'
			,message: 'Enter the file extension for module templates.'
			,default: this.rc.templateExtension || templateExtension
		}
	];

	this.prompt(prompts, function(props) {
		this.enableSprites = props.enableSprites;
		this.customModuleTemplatePath = props.customModuleTemplatePath || '';
		this.templateExtension = props.templateExtension || templateExtension;
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// SERVER CONFIG


XtcGenerator.prototype.sitename = function sitename() {
	var cb, prompts;

	if (!this.needServer) { return }
	cb = this.async();

	prompts = [
		{
			name: 'siteName'
			,message: 'Enter the website\'s name'
			,default: this.rc.siteName
		}
	];

	this.prompt(prompts, function(props) {
		this.siteName = props.siteName;
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Store setup config


XtcGenerator.prototype.storeRc = function storeRc() {
	this.rc.name                    = this.name;
	this.rc.repositoryUri           = this.repository.uri || null;
	this.rc.repositoryBranch        = this.repository.branch || null;
	this.rc.needServer              = this.needServer;
	this.rc.needBuild               = this.needBuild;
	this.rc.enableSprites           = this.enableSprites;
	this.rc.customModuleTemplatePath= this.customModuleTemplatePath || null;
	this.rc.templateExtension       = this.templateExtension || null;
	this.rc.siteName                = this.siteName || null;

	try {
		require('fs').writeFileSync('.xtcrc', JSON.stringify(this.rc, null, 2));
		console.log('\nCreated .xtcrc file\n');
	}
	catch (e) {
		console.error('\nUnable to write .xtcrc file.\nReason: %s\n', e.message);
	}
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Copy Files


XtcGenerator.prototype.copyFiles = function copyFiles() {

	// Project
	this.copy('package.json', projectPath('package.json'));
	this.copy('_.gitignore', projectPath('.gitignore'));
	//this.copy('_bower.json', 'bower.json');
	//this.copy('editorconfig', '.editorconfig');
	//this.copy('jshintrc', '.jshintrc');

	// Config
	this.mkdir(projectPath('_config'));
	this.copy('_config/configs.json', projectPath('_config/configs.json'));
	this.copy('_config/config-default.js', projectPath('_config/config-default.js'));
	this.copy('_config/config-local.json', projectPath('_config/config-local.json'));

	// README.md
	this.copy('_README.md', projectPath('README.md'));

	// Frontend
	this.directory('frontend-example', projectPath('frontend'));

	// Server
	if (this.needServer) {
		this.copy('_config/config-server.json', projectPath('_config/config-server.json'));
		this.copy('_config/config-secret.json', projectPath('_config/config-secret.json'));
		this.copy('_config/pinned-views.json', projectPath('_config/pinned-views.json'));
		this.directory('_controllers', projectPath('controllers'));

		// Helpers
		this.mkdir(projectPath('lib'));
		this.copy('_lib/handlebars-helpers.js', projectPath('lib/handlebars-helpers.js'));
		//this.copy('server.js', projectPath('server.js'));
	}

	// Build
	if (this.needBuild) {
		this.copy('Gruntfile.js', projectPath('Gruntfile.js'));
		this.copy('_config/config-build.json', projectPath('_config/config-build.json'));
	}

	// Deployment
	/*if (this.buildPackDeployment) {
		this.copy('Procfile', projectPath('Procfile'));
	}*/

};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Outro


//function onDependenciesInstalled() {
function onEnd() {

	var outro =
		'\nInstallation complete!\n\n' +

		chalk.cyan('xtc build')     +'\t\tstarts dev build\n' +
		chalk.cyan('xtc start')     +'\t\tstarts the server\n\n' +

		chalk.cyan('xtc help')      +'\t\tlist available commands\n\n'

		//chalk.cyan('npm run mkmod') +    '\t\tcreate a frontend module\n' +
		//chalk.cyan('npm run mkskin') +    '\t\tcreate a skin for a frontend module\n'
	;
	log(outro);
}


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Helpers


function projectPath(joinPath) {
	return path.join(_projectPath, joinPath);
}

function validateString(value) {
	return typeof value === 'string' && value.length > 0;
}

function validateGitUri(value) {
	return value === '' || /^https:\/\/[\w\.\:/\-~]+\.git$/.test(value);
}


function nl(count) {
	count = count || 1;

	while (count > 0) {
		console.log('');
		count--;
	}
}
