// generator-xtc/app/index.js
'use strict';

var util = require('util');
var path = require('path');
var fs = require('fs');
var yeoman = require('yeoman-generator');
var chalk = require('chalk');
var _projectPath = '';
var log = console.log;
var dir = console.dir;
var rcVersion = 0;
var xtcfileName = 'xtcfile.json';


var XtcGenerator = module.exports = function XtcGenerator(args, options, config) {
	yeoman.generators.Base.apply(this, arguments);

	this.projectPath = process.cwd();
	this.xtcPath = path.join(process.cwd(), 'node_modules/xtc');
	var xtcPkg = require(path.join(this.xtcPath, 'package.json'));
	this.xtcVersion = xtcPkg.version;
	_projectPath = this.projectPath;

	// .xtcfile stores setup config for setting defaults when running setup again or updating
	var getRc = function(projectPath) {
		var rcData = {}
		;

		try {
			rcData = fs.readFileSync(path.join(projectPath, xtcfileName));
			rcData = JSON.parse(rcData);
		} catch (e) {
			if (e.code !== 'ENOENT') {
				console.error(e.message);
			}
		}

		// store current values from rc file.
		this.currentRcVersion = rcData.rcVersion || null
		this.currentXtcVersion = rcData.xtcVersion || null;

		rcData.rcVersion = rcVersion;
		rcData.xtcVersion = xtcPkg.version;

		return rcData;
	};
	this.rc = getRc.call(this, _projectPath);


	this.on('end', onEnd);

	/*this.on('end', function () {
		this.installDependencies({
			skipInstall: options['skip-install'],
			callback: onDependenciesInstalled.bind(this)
		});
	});*/

	var intro =
		chalk.cyan('\nxtc@%s\n\n') +
		chalk.magenta("    _/__    _  __  ._  _ _/_  _  _  _ ") + '\n' +
		chalk.magenta("  ></ /_   /_///_///_'/_ /   /_//_'/ /") + '\n' +
		chalk.magenta("          /     |/           _/       ") + '\n\n' +
		chalk.blue(' * express-terrific project generator *\n\n') +
		chalk.cyan('Please answer a few questions to set up your new project.\n') +
		chalk.cyan('All choices can be changed later by running `xtc setup` or editing the config.\n')
	;

	log(intro, this.xtcVersion);
};

util.inherits(XtcGenerator, yeoman.generators.Base);


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// NAME AND REPOSITORY


XtcGenerator.prototype.basics = function basics() {
	var cb = this.async()
		,prompts = [
		{
			name: 'name'
			,message: 'Enter a package name for the new project (no spaces)'
			,validate: validateString
			,default: this.rc.name
		},
		{
			name: 'repoUri'
			,message: '"Read-only" repository URI (https://host/path/to/project.git or blank)'
			,validate: validateGitUri
			,default: this.rc.repositoryUri
		}
	];

	this.repository = {};

	this.prompt(prompts, function(props) {
		this.name = props.name;
		this.repository.uri = props.repoUri;
		cb();
	}.bind(this));
};


XtcGenerator.prototype.gitBranch = function gitBranch() {
	var cb, prompts;

	if (!this.repository.uri) { return }
	cb = this.async();

	prompts = [
		{
			name: 'branch'
			,message: 'What branch should be used when linking to project source files'
			,default: this.rc.repositoryBranch || 'develop'
		}
	];

	this.prompt(prompts, function(props) {
		this.repository.branch = props.branch;
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// CHOOSE SERVER AND BUILD PARTS


XtcGenerator.prototype.parts = function parts() {
	var cb = this.async()
		,prompts = [
		{
			name: 'neededParts'
			,type: 'checkbox'
			,message: 'What parts of xtc do you need'
			,choices: [
				{
					name: 'Server'
					,value: 'needServer'
					,checked: typeof this.rc.needServer == 'boolean' ? this.rc.needServer : true
				}
				,{
					name: 'Build'
					,value: 'needBuild'
					,checked: typeof this.rc.needBuild == 'boolean' ? this.rc.needBuild : true
				}
			]
		}
	];

	this.needBuild = false;
	this.needServer = false;
	this.configs = ['"default"'];

	this.prompt(prompts, function(props) {
		var i;
		for (i = 0; i < props.neededParts.length; i++) {
			var value = props.neededParts[i];
			this[value] = true;
		}

		this.needBuild && this.configs.push('"build"');
		this.needServer && this.configs.push('"server"', '"secret"');
		this.configs.push('"local"');
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// BUILD CONFIG


XtcGenerator.prototype.buildConfig = function buildConfig() {
	var cb, prompts
		,_projectPath = this.projectPath
		,isDir = false
		,templateExtension = '.hbs'
	;

	if (!this.needBuild) { return }
	cb = this.async();

	prompts = [
		{
			name    : 'enableSprites'
			,type   : 'confirm'
			,message: 'Enable sprites generation? Note: Needs a working "glue" (Python) installation.'
			,default: typeof this.rc.enableSprites === 'boolean' ? this.rc.enableSprites : false
		}
		,{
			name    : 'extendedCustomization'
			,type   : 'confirm'
			,message: 'Do you want to further customize the setup?'
			,default: this.rc.extendedCustomization || false
		}
		,{
			name    : 'customModuleTemplate'
			,when   : function(answers) { return answers.extendedCustomization }
			,type   : 'confirm'
			,message: 'Use a custom module template for the module generator?'
			,default: typeof this.rc.customModuleTemplatePath === 'string'
		}
		,{
			name    : 'customModuleTemplatePath'
			,when   : function(answers) { return answers.customModuleTemplate }
			,type   : 'input'
			,message: 'Template path for custom module generator. File (custom template) or directory (custom module).\n*'
			,default: this.rc.customModuleTemplatePath
			,validate: function(templatePath) {
				try {
					var stat = fs.lstatSync(path.resolve(_projectPath, templatePath));
				}
				catch (e) {
					return 'Path does not exist';
				}

				if (stat.isFile()) {
					templateExtension = path.extname(templatePath);
					return true;
				}
				if (stat.isDirectory()) {
					isDir = true;
					return true;
				}
			}
			,filter : function(templatePath) {
				// For now we're just gonna assume the standard config path! Should probably add a prompt for this too though
				return path.relative(_projectPath, templatePath);
			}
		}
		,{
			name    : 'templateExtension'
			,when   : function(answers) { return isDir }
			,type   : 'input'
			,message: 'Enter the file extension for module template files.'
			,default: this.rc.templateExtension || templateExtension
		}
	];

	this.prompt(prompts, function(props) {
		this.enableSprites              = props.enableSprites;
		this.extendedCustomization      = props.extendedCustomization;
		this.customModuleTemplatePath   = props.customModuleTemplatePath || '';
		this.templateExtension          = props.templateExtension || templateExtension;
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// SERVER CONFIG


XtcGenerator.prototype.sitename = function sitename() {
	var cb, prompts;

	if (!this.needServer) { return }
	cb = this.async();

	prompts = [
		{
			name: 'siteName'
			,message: 'Enter the website\'s name'
			,default: this.rc.siteName
		}
	];

	this.prompt(prompts, function(props) {
		this.siteName = props.siteName;
		cb();
	}.bind(this));
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Store setup config


XtcGenerator.prototype.storeRc = function storeRc() {
	this.rc.rcVersion               = this.rcVersion // already set, but here for completeness
	this.rc.xtcVersion              = this.xtcVersion // already set, but here for completeness
	this.rc.name                    = this.name;
	this.rc.repositoryUri           = this.repository.uri || null;
	this.rc.repositoryBranch        = this.repository.branch || null;
	this.rc.needServer              = this.needServer;
	this.rc.needBuild               = this.needBuild;
	this.rc.enableSprites           = this.enableSprites;
	this.rc.extendedCustomization   = this.extendedCustomization;
	this.rc.customModuleTemplatePath= this.customModuleTemplatePath || null;
	this.rc.templateExtension       = this.templateExtension || null;
	this.rc.siteName                = this.siteName || null;

	try {
		console.log('\nwriting %s\n', xtcfileName);
		fs.writeFileSync(xtcfileName, JSON.stringify(this.rc, null, 2));
	}
	catch (e) {
		console.error('\nUnable to write %s.\nReason: %s\n', xtcfileName, e.message);
	}
};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Copy Files


XtcGenerator.prototype.copyFiles = function copyFiles() {

	// Project
	this.copy('_package.json', projectPath('package.json'));
	this.copy('_.gitignore', projectPath('.gitignore'));
	//this.copy('_bower.json', 'bower.json');
	//this.copy('jshintrc', '.jshintrc');

	// Config
	this.mkdir(projectPath('config'));
	this.copy('_config/_.gitignore', projectPath('config/.gitignore'));
	this.copy('_config/configs.json', projectPath('config/configs.json'));
	this.copy('_config/config-default.js', projectPath('config/config-default.js'));
	this.copy('_config/config-local.json', projectPath('config/config-local.json'));

	// Frontend
	this.mkdir(projectPath('frontend'));
	this.directory('frontend-example/application', projectPath('frontend/application'));
	// also copies frontend-example/base/css/sprites/.gitignore
	this.directory('frontend-example/base', projectPath('frontend/base'));
	this.directory('frontend-example/inline', projectPath('frontend/inline'));
	this.directory('frontend-example/modules', projectPath('frontend/modules'));
	this.directory('frontend-example/public', projectPath('frontend/public'));

	// Server
	if (this.needServer) {
		this.directory('frontend-example/_views', projectPath('frontend/_views'));

		this.copy('server.js', projectPath('server.js'));
		this.copy('_config/config-server.json', projectPath('config/config-server.json'));
		this.copy('_config/config-secret.json', projectPath('config/config-secret.json'));
		this.copy('_config/pinned-views.json', projectPath('config/pinned-views.json'));
		this.directory('_controllers', projectPath('controllers'));

		// Helpers
		this.mkdir(projectPath('lib'));
		this.copy('lib/helpers.js', projectPath('lib/helpers.js'));
		this.copy('lib/handlebars-helpers.js', projectPath('lib/handlebars-helpers.js'));

		// README.md
		this.copy('_README.md', projectPath('README.md'));

		// Dot-files
		this.copy('editorconfig', '.editorconfig');
	}

	// Build
	if (this.needBuild) {
		this.copy('Gruntfile.js', projectPath('Gruntfile.js'));
		this.copy('_config/config-build.json', projectPath('config/config-build.json'));

		// Dot-files
		this.copy('editorconfig', 'frontend/.editorconfig');
	}

	// Docs
	this.copy(path.join(this.xtcPath, 'Documentation.md'), projectPath('docs/xtc-documentation.md'));

	// Deployment
	/*if (this.buildPackDeployment) {
		this.copy('Procfile', projectPath('Procfile'));
	}*/

};


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Outro


function onEnd() {}


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Helpers


function projectPath(joinPath) {
	return path.join(_projectPath, joinPath);
}

function validateString(value) {
	return typeof value === 'string' && value.length > 0;
}

function validateGitUri(value) {
	return value === '' || /^https:\/\/[\w\.\:/\-~]+\.git$/.test(value);
}


function nl(count) {
	count = count || 1;

	while (count > 0) {
		console.log('');
		count--;
	}
}
